### PR TITLE
Feat/ollama via LiteLLM

### PR DESCRIPTION
**Pull request by jaelin215**
# Pull Request

## Title

Feat/Ollama integration via Google ADK's LiteLLM wrapper

## Issue(s)

Fixes  #40 and #41 

## Summary

To enable free LLM model access, integrated Ollama to the existing code base. For persona and daily schedule generation, Ollama is called directly. For simulating ORPA/ORPDA, Ollama is called via Google ADK's LiteLLM wrapper. With current code, `config.yaml` remains a centralized configuration file that saves cognitive load of remembering where in which code file needs to be edited before running the code.

## Scope
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Chore (refactoring code, technical debt, workflow improvements)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Refactor (does not change functionality, e.g. code style improvements, linting)
- [x] Documentation update (e.g. README.md)

## Changes

- Added Ollama sample runner `app/src/utils/run_ollama_sample.py`
- Added LiteLLM sample runner `app/src/utils/run_litellm_sample.py`
- Added OpenRouter sample runner `app/src/utils/run_ollama_sample.py`
- Added `call_ollama` function to a new file `app/src/ollama_api.py` that can be used in `app/src/run_persona_injector.py`
- Renamed deprecated `call_gemini` API file to `app/src/orpda_runner_gemini.py` for future use
- Updated `app/src/orpda_runner.py`
  - Replaced Gemini model call with Ollama model call via Google ADK's LiteLLM
  - Added MODEL_NAME tag to Langfuge tracing (this block is used. confirmed in Langfuse)
- Updated `app/src/simulate.py` 
  - Added MODEL_NAME tag to Langfuge tracing (this block of code doesn't seem to be used. future: delete after verifying)
  - Added MODEL_NAME as suffix to log files. For example:
    - `app/logs/memory_streams_session_orpda_20260125_102705_deepseek-v3.2:cloud.log`
    - `app/logs/session_orpda_20260125_102705_deepseek-v3.2:cloud.log`
  - Added MODEL_NAME within the time stamp of each log for ease of log analysis in the future.
- Generated multiple driftville_personas.json file using multiple open source LLM models
  - llama: "llama2:latest"
  - mistral: "mistral:latest"
  - glm: "glm-4.7:cloud"
  - qwen: "qwen3-next:80b-cloud" # better than qwen3:8b
  - gpt-oss: "gpt-oss:20b-cloud"
  - deepseek: "deepseek-v3.2:cloud" # better than deepseek-r1:8b
  - gemini: "gemini-3-flash-preview:cloud"
- set the file `driftville_personas.json` the same as `app/src/driftville_personas_v12_deepseek-v3.2:cloud.json`
- Fixed PORT issues in `app2/app2.py`

## Acceptance Criteria

- [x] new driftville_personas.json is created using a free open source LLM model
- [x] new session logs are generated using a free open source LLM model (w/ MODEL_NAME included in sfuffix and timestamp)
- [x] new memory_streams logs are generated using a free open source LLM model (w/ MODEL_NAME included in sfuffix and timestamp)

## How I Tested

- Unit tests: `poetry run pytest -q`
- Lint/format: `poetry run isort . && poetry run black . && poetry run flake8 .`
- Manual checks / screenshots noted below.

**Prerequisite:**
In `app/config/config.yaml`, set default Ollama model in config

Screenshot 2026-01-25 at 11 52 29 AM

**Test generating persona and schedule [PASS]:**
1-1. Run persona and schedule generator
`python app/src/run_persona_injector.py`

1-2. Expected result:
`app/src/driftville_personas.json` is generated or updated.

**Test simulation [PASS]:**
2-1. In `app/config/config.yaml`, set configuration for a quick test (num_ticks=2, use_drift=True, load_prompt_from_langfuse=False)

Screenshot 2026-01-25 at 11 53 46 AM

[NOTE: also tested with 68 ticks. This long running simulation also didn't hit rate limits. PASS]

2-2. Run simulation
`python app/src/simulate.py`

2-3. Expected result:
- memory_streams log file is generated with correct MODEL_NAME suffix and timestamp (e.g., app/logs/memory_streams_session_orpda_20260125_102705_deepseek-v3.2:cloud.log)
- session log file is generated with correct MODEL_NAME suffix and timestamp (e.g., app/logs/session_orpda_20260125_102705_deepseek-v3.2:cloud.log)

## Screenshots / Artifacts (optional)

MODEL_NAME (i.e. deepseek) is captured correctly as a tag in Langfuse.

Screenshot 2026-01-25 at 12 02 28 PM

## API / DB Impact
- **API changes**: 
- **DB changes**: 

## Follow-ups / Dependencies

- TO-DO: Add model temperature control to Ollama/LiteLLM model
- TO-DO: Analyze logs to compare the content of drift between models (e.g., variability, frequency, intensity, etc.)

## Checklist
- [x] Code compiles and passes CI locally (format → quality → tests)
- [x] Added/updated tests for new logic and edge cases
- [x] Updated docs / README where relevant
- [x] No secrets or credentials committed

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -18,12 +18,14 @@ seaborn = "^0.13.2"
...
 google-adk = "^1.21.0"
 flask = "^3.1.2"
+ollama = "^0.6.1"
 
 [tool.poetry.group.dev.dependencies]
 isort = "^5.13.2"
 black = "^24.10.0"
 flake8 = "^7.1.1"
 pytest = "^8.3.3"
+litellm = {version = "<1.81.1", extras = ["proxy"]}
```
[the remark this answers is not in the document]
Comment (jaelin215) [reply pyproject.toml:28]: litellm=1.81.3 is required for langfuse OTEL.

File: app/config/config.py [hunk 2]
```diff
@@ -17,15 +17,52 @@ def load_config():
...
+# LLAMA = OLLAMA_MODELS.get("llama", "ollama/llama2:latest")
+# MISTRAL = OLLAMA_MODELS.get("mistral", "ollama/mistral:latest")
+# QWEN = OLLAMA_MODELS.get("qwen", "ollama/qwen3:8b")
+# GPT_OSS = OLLAMA_MODELS.get("gpt-oss", "ollama/gpt-oss:20b-cloud")
+# DEEPSEEK = OLLAMA_MODELS.get("deepseek", "ollama/deepseek-r1:8b")
+
+###############################
+# OpenRouter config
+###############################
+# OPENROUTER_MODELS = config.get("openrouter_models", {})
```
[the remark this answers is not in the document]
Comment (jaelin215) [reply app/config/config.py:68]: We will keep this here for a potential future use.

File: app/src/utils/run_ollama_sample.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+client = Client()
+messages = [
+    {
+        "role": "user",
+        "content": "what is DMN neuroscience?",
+    },
+]
+
+if use_stream:
+    for part in client.chat("gpt-oss:120b-cloud", messages=messages, stream=use_stream):
```
[the remark this answers is not in the document]
Comment (jaelin215) [reply app/src/utils/run_ollama_sample.py:19]: this works.

File: app/src/ollama_api.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+import time
+from collections import deque
+from pathlib import Path
+
+from dotenv import load_dotenv
+from ollama import Client
+
+# Paths
+ROOT = Path(__file__).resolve().parents[2]
+print(ROOT)
```
[the remark this answers is not in the document]
Comment (jaelin215) [reply app/src/ollama_api.py:17]: this is okay.

File: app/src/ollama_api.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        # Remove calls older than 1 minute
+        while self.calls and self.calls[0] < now - 60:
+            self.calls.popleft()
+
+        if len(self.calls) >= self.rate:
+            wait_time = 60 - (now - self.calls[0])
+            print(f"  [Rate limit: waiting {wait_time:.1f}s]")
+            await asyncio.sleep(wait_time + 0.1)
+
+        self.calls.append(time.time())
```
[the remark this answers is not in the document]
Comment (jaelin215) [reply app/src/ollama_api.py:43]: Update this in the future.

File: app/src/ollama_api.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+async def call_ollama(prompt, model=None, temperature=None, use_stream=False):
+    """Call any LLM using Ollama with rate limiting."""
+    await rate_limiter.acquire()
+
+    # Get model name and temperature from config, with fallbacks
+    final_model_name = model or MODEL_NAME
+    final_temperature = temperature or MODEL_TEMPERATURE
+    start_time = time.time()
+    client = Client()
```
[the remark this answers is not in the document]
Comment (jaelin215) [reply app/src/ollama_api.py:57]: Address this in the future.

File: app/src/orpda_runner.py [hunk 7]
```diff
@@ -31,9 +36,10 @@
 # Load environment & paths
 # -------------------------
 REPO_ROOT = Path(__file__).resolve().parents[2]
+print(REPO_ROOT)
 if str(REPO_ROOT) not in sys.path:
```
[the remark this answers is not in the document]
Comment (jaelin215) [reply app/src/orpda_runner.py:40]: this is okay for now.

File: app/src/orpda_runner.py [hunk 8]
```diff
@@ -43,7 +49,14 @@
...
 
-# from app.src.observe_non_llm_agent import deterministic_observe
+# Initialize local/cloud Ollama model via LiteLLM
+try:
+    my_local_model = LiteLlm(model=f"ollama/{MODEL_NAME}")
+    print("Successfully loaded a model via LiteLLM")
+except Exception as e:
+    print(f"Error: {type(e).__name__}: e")
+    print("Failed to load model vis LiteLLM")
+print("import complete")
```
[the remark this answers is not in the document]
Comment (jaelin215) [reply app/src/orpda_runner.py:60]: fixed

File: app/src/run_persona_injector.py [hunk 9]
```diff
@@ -85,20 +93,20 @@ def main() -> None:
...
+        help=f"Override model name (default: value in persona_injector.yaml, fallback {MODEL_NAME})",
     )
     args = parser.parse_args()
 
     cfg = load_prompt_config(YAML_PATH)
     instruction = cfg.get("instruction", "")
-    model_name = args.model or cfg.get("model") or DEFAULT_MODEL
-    os.environ["MODEL_NAME"] = model_name  # hint for gemini_api config
+    # model_name = args.model or cfg.get("model") or MODEL_NAME
+    # os.environ["MODEL_NAME"] = model_name  # hint for gemini_api config
```
[the remark this answers is not in the document]
Comment (jaelin215) [reply app/src/run_persona_injector.py]: removed

File: app/src/run_persona_injector.py [hunk 10]
```diff
@@ -85,20 +93,20 @@ def main() -> None:
...
 
     persona_text = args.input.read_text(encoding="utf-8")
     try:
         raw = asyncio.run(
             asyncio.wait_for(
-                call_llm(instruction, persona_text, model_name), timeout=60
+                call_llm(instruction, persona_text, MODEL_NAME), timeout=60
             )
```
[the remark this answers is not in the document]
Comment (jaelin215) [reply app/src/run_persona_injector.py:108]: updated to 180seconds (3min).

**Review by jaelin215 — COMMENTED** [review]
Reviewed and applied copilot suggested changes.

File: pyproject.toml [hunk 11]
```diff
@@ -9,7 +9,7 @@ package-mode = false
 
 [tool.poetry.dependencies]
 python = "^3.12"
-langfuse = "^3.10.6"
+langfuse = "3.12.0"
```
[the remark this answers is not in the document]
Comment (jaelin215) [reply pyproject.toml:12]: langfuse 3.12.0 was needed for compatibility with litellm to enable Langfuse OTEL.
